In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
from google.colab import files
data=pd.read_csv('StudentsPerformance.csv')

In [ ]:
data.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

#第一題

In [ ]:
f=data.groupby('gender')
frequence=f.size()
flist=pd.DataFrame(frequence,columns=['Frequence_性別'])
f=data.groupby('gender')
percentage=f.size()/len(data)
plist=pd.DataFrame(percentage,columns=["Percentage_性別"])
flist.merge(plist,on = 'gender')

Frequence_性別  Percentage_性別
gender                             
female           518          0.518
male             482          0.482

在性別這筆資料當中

女生518人

男生482人

比例為0.518:0.482

In [ ]:
f=data.groupby('race/ethnicity')
frequence=f.size()
flist=pd.DataFrame(frequence,columns=['Frequence_組別'])
f=data.groupby('race/ethnicity')
percentage=f.size()/len(data)
plist=pd.DataFrame(percentage,columns=["Percentage_組別"])
flist.merge(plist,on = 'race/ethnicity')

Frequence_組別  Percentage_組別
race/ethnicity                             
group A                   89          0.089
group B                  190          0.190
group C                  319          0.319
group D                  262          0.262
group E                  140          0.140

在組別這筆資料當中

A組89人

B組190人

C組319人

D組262人

E組140人

比例為0.089:0.190:0.319:0.262:0.140

In [ ]:
f=data.groupby('parental level of education')
frequence=f.size()
flist=pd.DataFrame(frequence,columns=['Frequence_家長教育程度'])
percentage=f.size()/len(data)
plist=pd.DataFrame(percentage,columns=["Percentage_家長教育程度"])
flist.merge(plist,on = 'parental level of education')

Frequence_家長教育程度  Percentage_家長教育程度
parental level of education                                     
associate's degree                        222              0.222
bachelor's degree                         118              0.118
high school                               196              0.196
master's degree                            59              0.059
some college                              226              0.226
some high school                          179              0.179

在家長教育程度資料當中

副學士學位222人

學士學位118人

高中196人

碩士學位59人

五專226人

高職179人

比例為0.222:0.118:0.196:0.059:0.226:0.179

In [ ]:
f=data.groupby('lunch')
frequence=f.size()
flist=pd.DataFrame(frequence,columns=['Frequence_午餐費用'])
percentage=f.size()/len(data)
plist=pd.DataFrame(percentage,columns=["Percentage_午餐費用"])
flist.merge(plist,on = 'lunch')

Frequence_午餐費用  Percentage_午餐費用
lunch                                        
free/reduced             355            0.355
standard                 645            0.645

在午餐費用資料當中

免費/較便宜355人

標準645人

比例為0.355:0.645

In [ ]:
f=data.groupby('test preparation course')
frequence=f.size()
flist=pd.DataFrame(frequence,columns=['Frequence_考試準備課程'])
percentage=f.size()/len(data)
plist=pd.DataFrame(percentage,columns=["Percentage_考試準備課程"])
flist.merge(plist,on = 'test preparation course')

Frequence_考試準備課程  Percentage_考試準備課程
test preparation course                                     
completed                             358              0.358
none                                  642              0.642

在考試準備課程資料當中

完成358人

沒完成642人

比例為0.358:0.642

In [ ]:
#描述性分析
data.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

三科都有1000筆資料，最大值都為100

數學平均66.09，閱讀平均69.17，寫作平均68.05

數學標準差15.16，閱讀標準差14.60，寫作標準差15.20

數學最小值0，閱讀最小值17，寫作最小值10

數學25百分位57，閱讀25百分位59，寫作25百分位57.75

數學50百分位66，閱讀50百分位70，寫作50百分位69

數學75百分位77，閱讀75百分位79，寫作75百分位79

#第二題

不同性別的成績差異

In [ ]:
from scipy import stats
(t,p)=stats.ttest_ind(data['math score'][data['gender']=='female'],data['math score'][data['gender']=='male'])
print('數學 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['reading score'][data['gender']=='female'],data['reading score'][data['gender']=='male'])
print('閱讀 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['writing score'][data['gender']=='female'],data['writing score'][data['gender']=='male'])
print('寫作 t value= %.2f,  p_value= %.2f'%(t,p))

數學 t value= -5.38,  p_value= 0.00
閱讀 t value= 7.96,  p_value= 0.00
寫作 t value= 9.98,  p_value= 0.00


因為p值小於0.05，所以不同性別的數學/閱讀/寫作成績有顯著差異

In [ ]:
a=data.groupby('gender').mean()
b=a[["math score"]]
b

math score
gender            
female   63.633205
male     68.728216

女生數學平均63.633

男生數學平均68.728

In [ ]:
a=data.groupby('gender').mean()
b=a[["reading score"]]
b

reading score
gender               
female      72.608108
male        65.473029

女生閱讀平均72.608

男生閱讀平均65.473

In [ ]:
a=data.groupby('gender').mean()
b=a[["writing score"]]
b

writing score
gender               
female      72.467181
male        63.311203

女生寫作平均72.467

男生寫作平均63.311

不同組別的成績差異

In [ ]:
from scipy import stats
(f,pv)=stats.f_oneway(data['math score'][data['race/ethnicity']=='group A'],
                          data['math score'][data['race/ethnicity']=='group B'],
                          data['math score'][data['race/ethnicity']=='group C'],
                          data['math score'][data['race/ethnicity']=='group D'],
                          data['math score'][data['race/ethnicity']=='group E'])
print('數學 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))
(f,pv)=stats.f_oneway(data['reading score'][data['race/ethnicity']=='group A'],
                          data['reading score'][data['race/ethnicity']=='group B'],
                          data['reading score'][data['race/ethnicity']=='group C'],
                          data['reading score'][data['race/ethnicity']=='group D'],
                          data['reading score'][data['race/ethnicity']=='group E'])
print('閱讀 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))
(f,pv)=stats.f_oneway(data['writing score'][data['race/ethnicity']=='group A'],
                          data['writing score'][data['race/ethnicity']=='group B'],
                          data['writing score'][data['race/ethnicity']=='group C'],
                          data['writing score'][data['race/ethnicity']=='group D'],
                          data['writing score'][data['race/ethnicity']=='group E'])
print('寫作 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))

數學 onewayANOVA=14.59, p-value=0.00
閱讀 onewayANOVA=5.62, p-value=0.00
寫作 onewayANOVA=7.16, p-value=0.00


因為p值小於0.05，所以不同組別的數學/閱讀/寫作成績有顯著差異

In [ ]:
a=data.groupby('race/ethnicity').mean()
b=a[["math score"]]
b

math score
race/ethnicity            
group A          61.629213
group B          63.452632
group C          64.463950
group D          67.362595
group E          73.821429

A組數學平均61.629

B組數學平均63.453

C組數學平均64.464

D組數學平均67.362

E組數學平均73.821

In [ ]:
a=data.groupby('race/ethnicity').mean()
b=a[["reading score"]]
b

reading score
race/ethnicity               
group A             64.674157
group B             67.352632
group C             69.103448
group D             70.030534
group E             73.028571

A組閱讀平均64.674

B組閱讀平均67.352

C組閱讀平均69.103

D組閱讀平均70.030

E組閱讀平均73.029

In [ ]:
a=data.groupby('race/ethnicity').mean()
b=a[["writing score"]]
b

writing score
race/ethnicity               
group A             62.674157
group B             65.600000
group C             67.827586
group D             70.145038
group E             71.407143

A組寫作平均62.674

B組寫作平均65.600

C組寫作平均67.827

D組寫作平均70.145

E組寫作平均71.407

不同家長教育程度的成績差異

In [ ]:
from scipy import stats
(f,pv)=stats.f_oneway(data['math score'][data['parental level of education']=="bachelor's degree"],
                          data['math score'][data['parental level of education']=='some college'],
                          data['math score'][data['parental level of education']=="master's degree"],
                          data['math score'][data['parental level of education']=="associate's degree"],
                          data['math score'][data['parental level of education']=='some high school'],
                          data['math score'][data['parental level of education']=='high school'])
print('數學 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))
(f,pv)=stats.f_oneway(data['reading score'][data['parental level of education']=="bachelor's degree"],
                          data['reading score'][data['parental level of education']=='some college'],
                          data['reading score'][data['parental level of education']=="master's degree"],
                          data['reading score'][data['parental level of education']=="associate's degree"],
                          data['reading score'][data['parental level of education']=='some high school'],
                          data['reading score'][data['parental level of education']=='high school'])
print('閱讀 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))
(f,pv)=stats.f_oneway(data['writing score'][data['parental level of education']=="bachelor's degree"],
                          data['writing score'][data['parental level of education']=='some college'],
                          data['writing score'][data['parental level of education']=="master's degree"],
                          data['writing score'][data['parental level of education']=="associate's degree"],
                          data['writing score'][data['parental level of education']=='some high school'],
                          data['writing score'][data['parental level of education']=='high school'])
print('寫作 onewayANOVA=%.2f, p-value=%.2f'%(f,pv))

數學 onewayANOVA=6.52, p-value=0.00
閱讀 onewayANOVA=9.29, p-value=0.00
寫作 onewayANOVA=14.44, p-value=0.00


因為p值小於0.05，所以不同家長教育程度的數學/閱讀/寫作成績有顯著差異

In [ ]:
a=data.groupby('parental level of education').mean()
b=a[["math score"]]
b

math score
parental level of education            
associate's degree            67.882883
bachelor's degree             69.389831
high school                   62.137755
master's degree               69.745763
some college                  67.128319
some high school              63.497207

副學士學位數學平均67.883

學士學位數學平均69.390

高中數學平均62.138

碩士學位數學平均69.746

五專數學平均67.128

高職數學平均63.497

In [ ]:
a=data.groupby('parental level of education').mean()
b=a[["reading score"]]
b

reading score
parental level of education               
associate's degree               70.927928
bachelor's degree                73.000000
high school                      64.704082
master's degree                  75.372881
some college                     69.460177
some high school                 66.938547

副學士學位閱讀平均70.927

學士學位閱讀平均73.000

高中閱讀平均64.704

碩士學位閱讀平均75.372

五專閱讀平均69.460

高職閱讀平均66.939

In [ ]:
a=data.groupby('parental level of education').mean()
b=a[["writing score"]]
b

writing score
parental level of education               
associate's degree               69.896396
bachelor's degree                73.381356
high school                      62.448980
master's degree                  75.677966
some college                     68.840708
some high school                 64.888268

副學士學位寫作平均69.896

學士學位寫作平均73.381

高中寫作平均62.449

碩士學位寫作平均75.678

五專寫作平均68.841

高職寫作平均64.888

不同午餐費用的差異

In [ ]:
from scipy import stats
(t,p)=stats.ttest_ind(data['math score'][data['lunch']=='standard'],data['math score'][data['lunch']=='free/reduced'])
print('數學 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['reading score'][data['lunch']=='standard'],data['reading score'][data['lunch']=='free/reduced'])
print('閱讀 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['writing score'][data['lunch']=='standard'],data['writing score'][data['lunch']=='free/reduced'])
print('寫作 t value= %.2f,  p_value= %.2f'%(t,p))

數學 t value= 11.84,  p_value= 0.00
閱讀 t value= 7.45,  p_value= 0.00
寫作 t value= 8.01,  p_value= 0.00


因為p值小於0.05，所以不同午餐費用的數學/閱讀/寫作成績有顯著差異

In [ ]:
a=data.groupby('lunch').mean()
b=a[["math score"]]
b

math score
lunch                   
free/reduced   58.921127
standard       70.034109

免費/較便宜數學平均58.921

標準數學平均70.034

In [ ]:
a=data.groupby('lunch').mean()
b=a[["reading score"]]
b

reading score
lunch                      
free/reduced      64.653521
standard          71.654264

免費/較便宜閱讀平均64.654

標準閱讀平均71.654

In [ ]:
a=data.groupby('lunch').mean()
b=a[["writing score"]]
b

writing score
lunch                      
free/reduced      63.022535
standard          70.823256

免費/較便宜寫作平均63.023

標準寫作平均70.823

不同考試準備課程的差異

In [ ]:
from scipy import stats
(t,p)=stats.ttest_ind(data['math score'][data['test preparation course']=='none'],data['math score'][data['test preparation course']=='completed'])
print('數學 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['reading score'][data['test preparation course']=='none'],data['reading score'][data['test preparation course']=='completed'])
print('閱讀 t value= %.2f,  p_value= %.2f'%(t,p))
(t,p)=stats.ttest_ind(data['writing score'][data['test preparation course']=='none'],data['writing score'][data['test preparation course']=='completed'])
print('寫作 t value= %.2f,  p_value= %.2f'%(t,p))

數學 t value= -5.70,  p_value= 0.00
閱讀 t value= -7.87,  p_value= 0.00
寫作 t value= -10.41,  p_value= 0.00


因為p值小於0.05，所以不同考試準備課程的數學/閱讀/寫作成績有顯著差異

In [ ]:
a=data.groupby('test preparation course').mean()
b=a[["math score"]]
b

math score
test preparation course            
completed                 69.695531
none                      64.077882

完成數學平均69.700

未完成數學平均64.078

In [ ]:
a=data.groupby('test preparation course').mean()
b=a[["reading score"]]
b

reading score
test preparation course               
completed                    73.893855
none                         66.534268

完成閱讀平均73.894

未完成閱讀平均66.534

In [ ]:
a=data.groupby('test preparation course').mean()
b=a[["writing score"]]
b

writing score
test preparation course               
completed                    74.418994
none                         64.504673

完成寫作平均74.419

未完成寫作平均64.505

#第三題

In [ ]:
qu=['math score','reading score','writing score']
from scipy import stats
name=[]
alist=[]
pvlist=[]
for i in range(len(qu)):
  (a,pv)=stats.pearsonr(data['math score'],data[qu[i]]) 
  name.append(qu[i])
  alist.append(a)
  pvlist.append(pv)
pd.set_option('display.float_format',lambda x:'%.5f'%x)
result=pd.DataFrame(alist,name,columns=['相關係數'])
result.insert(1,'p值',pvlist) 
result

相關係數      p值
math score    1.00000 0.00000
reading score 0.81758 0.00000
writing score 0.80264 0.00000

In [ ]:
qu=['math score','reading score','writing score']
from scipy import stats
name=[]
alist=[]
pvlist=[]
for i in range(len(qu)):
  (a,pv)=stats.pearsonr(data['reading score'],data[qu[i]]) 
  name.append(qu[i])
  alist.append(a)
  pvlist.append(pv)
pd.set_option('display.float_format',lambda x:'%.5f'%x)
result=pd.DataFrame(alist,name,columns=['相關係數'])
result.insert(1,'p值',pvlist) 
result

相關係數      p值
math score    0.81758 0.00000
reading score 1.00000 0.00000
writing score 0.95460 0.00000

In [ ]:
qu=['math score','reading score','writing score']
from scipy import stats
name=[]
alist=[]
pvlist=[]
for i in range(len(qu)):
  (a,pv)=stats.pearsonr(data['writing score'],data[qu[i]]) 
  name.append(qu[i])
  alist.append(a)
  pvlist.append(pv)
pd.set_option('display.float_format',lambda x:'%.5f'%x)
result=pd.DataFrame(alist,name,columns=['相關係數'])
result.insert(1,'p值',pvlist) 
result

相關係數      p值
math score    0.80264 0.00000
reading score 0.95460 0.00000
writing score 1.00000 0.00000

第三題分析

因為p值小於0.05

所以三個成績有顯著差異

#綜合上述

因為p<0.05

所以不同的gender、race/ethnicity、parental level of education、lunch、test preparation和course對三個成績都有顯著差異，三個成績之間也有顯著差異

